File: ios-native/ClubFuoco/Features/Bookings/SavedEventsSection.swift
import SwiftUI

/// Events the guest saved but hasn't paid for, pinned above their real tickets.
///
/// These are NOT tickets and the section works hard not to look like one: no
/// QR, no Wallet button, a dashed edge instead of a solid card, and the price
/// stated as an action rather than a fact. A guest who mistakes a bookmark for
/// entry finds out at a door, which is the worst possible place.
struct SavedEventsSection: View {
    @Environment(\.api) private var api
    @State private var events: [SavedEvent] = []
    @State private var loading = true

    /// Opens the invite so they can pay.
    var onOpen: (String) -> Void

    var body: some View {
        Group {
            if !events.isEmpty {
                VStack(alignment: .leading, spacing: 12) {
                    Text("SAVED — NOT PAID YET")
                        .font(.cfMono(10)).kerning(2)
                        .foregroundStyle(Theme.gold)

                    ForEach(events) { e in
                        Button { onOpen(e.inviteToken) } label: { row(e) }
                            .buttonStyle(.plain)
                    }
                }
            }
        }
        .task { await load() }
        // A save made on the invite sheet should show up here without a relaunch.
        .onReceive(NotificationCenter.default.publisher(for: .cfInviteClaimed)) { _ in
            Task { await load() }
        }
    }

    private func row(_ e: SavedEvent) -> some View {
        HStack(spacing: 14) {
            VStack(alignment: .leading, spacing: 3) {
                Text(e.title ?? e.venueName)
                    .font(.cfSans(15, weight: .medium))
                    .foregroundStyle(Theme.parchment)
                    .lineLimit(1)
                Text("\(e.venueName) · \(SavedEvent.formatDate(e.nightDate))")
                    .font(.cfSans(12))
                    .foregroundStyle(Theme.parchment.opacity(0.6))
            }
            Spacer(minLength: 8)
            // The price reads as the thing left to do, not as a label.
            Text(e.priceCents > 0 ? "Pay \(e.priceLabel)" : "Join")
                .font(.cfSans(13, weight: .semibold))
                .foregroundStyle(Theme.emberCream)
                .padding(.horizontal, 14).padding(.vertical, 8)
                .background(Capsule().fill(Theme.ember))
        }
        .padding(14)
        .background(
            RoundedRectangle(cornerRadius: 14)
                .strokeBorder(style: StrokeStyle(lineWidth: 1, dash: [5, 4]))
                .foregroundStyle(Theme.gold.opacity(0.35))
        )
        .contentShape(Rectangle())
    }

    private func load() async {
        defer { loading = false }
        struct Resp: Decodable, Sendable { let events: [SavedEvent] }
        // Signed out is the common case here, not an error — stay silent.
        guard let resp: Resp = try? await api.get("/api/me/saved-events") else {
            events = []
            return
        }
        events = resp.events
    }
}

struct SavedEvent: Decodable, Identifiable, Sendable {
    let allocationId: String
    let inviteToken: String
    let title: String?
    let nightDate: String
    let venueName: String
    let priceCents: Int
    let currency: String?

    var id: String { allocationId }

    var priceLabel: String {
        let symbol = (currency ?? "eur").lowercased() == "eur" ? "€" : ""
        return priceCents % 100 == 0
            ? "\(symbol)\(priceCents / 100)"
            : String(format: "%@%.2f", symbol, Double(priceCents) / 100)
    }

    static func formatDate(_ iso: String) -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        guard let d = f.date(from: iso) else { return iso }
        let out = DateFormatter()
        out.dateFormat = "EEE d MMM"
        return out.string(from: d)
    }
}

File: android-native/app/src/main/java/com/clubfuoco/app/features/bookings/SavedEventsSection.kt
package com.clubfuoco.app.features.bookings

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.clubfuoco.app.R
import com.clubfuoco.app.core.designsystem.Geist
import com.clubfuoco.app.core.designsystem.GeistMono
import com.clubfuoco.app.core.designsystem.Theme
import com.clubfuoco.app.core.designsystem.clickableUnlessBusy
import com.clubfuoco.app.core.network.ApiClient
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * Events the guest bookmarked but has NOT paid for, pinned above their real
 * tickets.
 *
 * These are not tickets and the section works hard not to look like one: no QR,
 * a dashed edge instead of a solid card, and the price stated as the thing left
 * to do rather than as a fact. Someone who mistakes a bookmark for entry finds
 * out at a door, which is the worst possible place to find out.
 *
 * Port of `SavedEventsSection`.
 */
@Composable
fun SavedEventsSection(
    api: ApiClient,
    /** Opens the invite so they can pay. */
    onOpen: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    var events by remember { mutableStateOf<List<SavedEvent>>(emptyList()) }

    LaunchedEffect(Unit) {
        // Signed out is the common case here, not an error — stay silent.
        events = runCatching {
            withContext(Dispatchers.IO) {
                api.get("/api/me/saved-events", SavedEventsResponse.serializer())
            }
        }.getOrNull()?.events.orEmpty()
    }

    if (events.isEmpty()) return

    Column(modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(
            stringResource(R.string.saved_notPaidYet).uppercase(),
            fontFamily = GeistMono, fontSize = 10.sp, letterSpacing = 2.sp,
            color = Theme.gold,
        )
        events.forEach { event -> SavedRow(event) { onOpen(event.inviteToken) } }
    }
}

@Composable
private fun SavedRow(event: SavedEvent, onClick: () -> Unit) {
    val dash = Theme.gold.copy(alpha = 0.35f)

    Box(Modifier.fillMaxWidth().clickableUnlessBusy(onClick = onClick)) {
        // Dashed, not solid: the edge is the first thing that says "this is not
        // a ticket", before any of the words do.
        Canvas(Modifier.matchParentSize()) {
            drawRoundRect(
                color = dash,
                size = Size(size.width, size.height),
                cornerRadius = CornerRadius(14.dp.toPx()),
                style = Stroke(
                    width = 1.dp.toPx(),
                    pathEffect = PathEffect.dashPathEffect(
                        floatArrayOf(5.dp.toPx(), 4.dp.toPx()),
                    ),
                ),
            )
        }

        Row(
            Modifier.fillMaxWidth().padding(14.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(3.dp)) {
                Text(
                    event.title ?: event.venueName,
                    fontFamily = Geist, fontWeight = FontWeight.Medium, fontSize = 15.sp,
                    color = Theme.ink,
                    maxLines = 1, overflow = TextOverflow.Ellipsis,
                )
                Text(
                    "${event.venueName} · ${formatNight(event.nightDate)}",
                    fontFamily = Geist, fontSize = 12.sp, color = Theme.stone,
                    maxLines = 1, overflow = TextOverflow.Ellipsis,
                )
            }
            // Reads as the thing left to do, not as a label.
            Text(
                if (event.priceCents > 0) {
                    stringResource(R.string.saved_pay, event.priceLabel)
                } else {
                    stringResource(R.string.saved_join)
                },
                fontFamily = Geist, fontWeight = FontWeight.SemiBold, fontSize = 13.sp,
                color = Theme.cream,
                modifier = Modifier
                    .clip(CircleShape)
                    .background(Theme.accent)
                    .padding(horizontal = 14.dp, vertical = 8.dp),
            )
        }
    }
}

@Serializable
private data class SavedEventsResponse(val events: List<SavedEvent> = emptyList())

@Serializable
data class SavedEvent(
    @SerialName("allocationId") val allocationId: String,
    @SerialName("inviteToken") val inviteToken: String,
    val title: String? = null,
    @SerialName("nightDate") val nightDate: String = "",
    @SerialName("venueName") val venueName: String = "",
    @SerialName("priceCents") val priceCents: Int = 0,
    val currency: String? = null,
) {
    val priceLabel: String
        get() {
            val symbol = if ((currency ?: "eur").lowercase() == "eur") "€" else ""
            return if (priceCents % 100 == 0) "$symbol${priceCents / 100}"
            else String.format(Locale.US, "%s%.2f", symbol, priceCents / 100.0)
        }
}

private fun formatNight(ymd: String): String = runCatching {
    LocalDate.parse(ymd).format(DateTimeFormatter.ofPattern("EEE d MMM", Locale.getDefault()))
}.getOrElse { ymd }
